File: 0283_Move_Zeroes.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int n = nums.size(), i = 0, j, k;
        while (i < n) {
            j = i;
            while (nums[j] == 0 && j < n)
                j++;
            for (k = 0; k < j - i && j + k < n; k++) {
                nums[i + k] = nums[j + k];
                nums[j + k] = 0;
            }
            i++;
            // i = i == j ? i + 1 : j;
        }
        return;
    }
};

File: 0283_Move_Zeroes.py
class Solution:
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        non_zero = []
        for i, val in enumerate(nums):
            if val != 0:
                non_zero.append(i)
        i = 0
        for nz in non_zero:
            nums[i] = nums[nz]
            i += 1
        
        for j in range(i,len(nums)):
            nums[j] = 0

class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        j = 0
        for i in range(n):
            if nums[i] != 0:
                nums[i], nums[j] = nums[j], nums[i]
                j += 1
        return
